pkg/doc/signature/jsonld/processor_test.go

```diff
@@ -1607,3 +1665,222 @@ func TestTransformBlankNodes(t *testing.T) {
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoader(createInMemoryDocumentLoader("http://localhost:8652/dummy.jsonld", extraJSONLDContext)),
+				},
+			},
+			{
+				name:   "canonizing sample document with extra cached dummy context",
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoaderCache(createContextCache()),
+				},
+			},
+			{
+				name:   "canonizing sample document with extra cached dummy context and cached document loader",
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoaderCache(createContextCache()),
+					WithDocumentLoader(jsonld.NewDefaultCachingDocumentLoader()),
+				},
+			},
+			{
+				name:   "canonizing sample document with extra cached dummy context and non caching document loader",
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoaderCache(createContextCache()),
+					WithDocumentLoader(jsonld.NewDefaultCachingDocumentLoader()),
+				},
+			},
+			{
+				name:   "canonizing sample document with extra byte cached dummy context",
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoaderCache(map[string]interface{}{
+						"http://localhost:8652/dummy.jsonld": []byte(extraJSONLDContext),
+					}),
+				},
+			},
+			{
+				name:   "canonizing sample document with extra map cached dummy context",
+				doc:    jsonLDMultipleInvalidRDFs,
+				result: canonizedSampleVP_extraContext,
+				opts: []ProcessorOpts{
+					WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+					WithDocumentLoaderCache(map[string]interface{}{
+						"http://localhost:8652/dummy.jsonld": stringToMapBench(b, extraJSONLDContext),
+					}),
+				},
+			},
+			// {
+			//	name:   "canonizing sample document with extra io.Reader cached dummy context",
+			//	doc:    jsonLDMultipleInvalidRDFs,
+			//	result: canonizedSampleVP_extraContext,
+			//	opts: []ProcessorOpts{
+			//		WithRemoveAllInvalidRDF(), WithExternalContext("http://localhost:8652/dummy.jsonld"),
+			//		WithDocumentLoaderCache(map[string]interface{}{
+			//			"http://localhost:8652/dummy.jsonld": strings.NewReader(extraJSONLDContext),
+			//		}),
+			//	},
+			// },
```

Comment:
why it is commented out?

Reply:
Because the buffer in `strings.NewReader()` is not thread safe.
One would need to use a thread safe Reader implementation for this to work. Example: https://gist.github.com/jmackie/11570bdcd8a4c10d72619a5e1f21c5f8

I can probably add a note in the DocumentLoader constructor about this. But this is an edge case with a seldom use. Users need to be aware to avoid using strings.NewReader() as a sync.Map value.